src/Client/Updater.php

```diff
@@ -398,31 +420,36 @@ private function updateRoot(RootMetadata &$rootData)
-                throw new \Exception("The maximum number root files have already been dowloaded:" . static::MAX_ROOT_DOWNLOADS);
+        try {
+            while ($nextRootContents = $this->repoFileFetcher->fetchFile("$nextVersion.root.json", static::MAXIMUM_DOWNLOAD_BYTES)) {
+                $rootsDownloaded++;
+                if ($rootsDownloaded > static::MAX_ROOT_DOWNLOADS) {
+                    throw new \Exception("The maximum number root files have already been dowloaded:" . static::MAX_ROOT_DOWNLOADS);
+                }
+                $nextRoot = RootMetadata::createFromJson($nextRootContents);
+                // *TUF-SPEC-v1.0.9 Section 5.1.3
+                $this->checkSignatures($nextRoot);
+                // Update Role and Key databases to use the new root information.
+                $this->roleDB = RoleDB::createFromRootMetadata($nextRoot);
+                $this->keyDB = KeyDB::createFromRootMetadata($nextRoot);
+                $this->checkSignatures($nextRoot);
+                // *TUF-SPEC-v1.0.9 Section 5.1.4
+                static::checkRollbackAttack($rootData, $nextRoot, $nextVersion);
+                $rootData = $nextRoot;
+                // *TUF-SPEC-v1.0.9 Section 5.1.5 - Needs no action.
+                // Note that the expiration of the new (intermediate) root metadata
+                // file does not matter yet, because we will check for it in step
+                // 1.8.
+
+                // *TUF-SPEC-v1.0.9 Section 5.1.6 and 5.1.7
+                $this->durableStorage['root.json'] = $nextRootContents;
+                $nextVersion = $rootData->getVersion() + 1;
+                // *TUF-SPEC-v1.0.9 Section 5.1.8 Repeat the above steps.
             }
-            $nextRoot = RootMetadata::createFromJson($nextRootContents);
-            // *TUF-SPEC-v1.0.9 Section 5.1.3
-            $this->checkSignatures($nextRoot);
-            // Update Role and Key databases to use the new root information.
-            $this->roleDB = RoleDB::createFromRootMetadata($nextRoot);
-            $this->keyDB = KeyDB::createFromRootMetadata($nextRoot);
-            $this->checkSignatures($nextRoot);
-            // *TUF-SPEC-v1.0.9 Section 5.1.4
-            static::checkRollbackAttack($rootData, $nextRoot, $nextVersion);
-            $rootData = $nextRoot;
-            // *TUF-SPEC-v1.0.9 Section 5.1.5 - Needs no action.
-            // Note that the expiration of the new (intermediate) root metadata
-            // file does not matter yet, because we will check for it in step
-            // 1.8.
-
-            // *TUF-SPEC-v1.0.9 Section 5.1.6 and 5.1.7
-            $this->durableStorage['root.json'] = $nextRootContents;
-            $nextVersion = $rootData->getVersion() + 1;
-            // *TUF-SPEC-v1.0.9 Section 5.1.8 Repeat the above steps.
+        } catch (RepoFileNotFound $exception) {
+            // This exception is expected when the newest root file has already been downloaded.
```

Comment:
I'm not sure I would do this; it feels like a misuse of an exception, and it might cover up some sort of freak error condition (maybe the file DOES exist, but the server is returning a 404 due to some error condition). IMHO we should just add a `fileExists()` method to `RepoFileFetcherInterface` to handle this case, so we don't try to download a non-existent file as a way of verifying its existence in the first place.

Reply:
Sorry, not sure I understand. How else would you check whether the file exist or not on the repo?

Reply:
In the case of the GuzzleFileFetcher implemented by #81, I was thinking maybe a `HEAD` request would do the trick, instead of `GET` (which will be heavier because it'll return the full content of the file). Using a lighter-weight request in order to check existence also gives us some wiggle room to automatically retry a few times before giving up.

On the other hand, I could be overthinking it too :)

Reply:
I think a HEAD request, and then, if successful, followed by a GET request makes sense 🙂

Reply:
@phenaproxima hadn't seen your last comment but was thinking of adding 
`/**
     * Gets a file if it exists if it exists in the remote repo.
     *
     * @param string $fileName
     *   The file name to fetch.
     *
     * @return string|null
     *   The contents of the file or null if it does not exist.
     */
    private function fetchFileIfExists(string $fileName):?string
    {
        try {
            return $this->repoFileFetcher->fetchFile($fileName, static::MAXIMUM_DOWNLOAD_BYTES);
        } catch (RepoFileNotFound $exception) {
            return null;
        }
    }`

Then we would need the try/catch around the loop. This was the `Updater` class but could be for `RepoFileFetcherInterface`

Reply:
add to RepoFileFetcherInterface